### ISSUE TITLE
Model is SO(3) equivariant only for the first 12 testing elements  

### ISSUE THREAD

**Issue by JacobHelwig** [QUVA-Lab/escnn#22]
Hello,

I would like to construct an SO(3) equivariant model that takes a vector field as input and outputs a vector field. My attempt is equivariant to the first 12 elements of `rot3donR3.testing_elements`, but fails for the remaining 48.

I think it is because of the field type for the hidden layers. I am mapping the input vector field to a scalar field, and then back again. I also tried using `irrep(j)`, `j=1,2,3` for the hidden feature type after removing the normalization and activation functions but still passed/ failed for the same elements as when using the trivial representation.

Is there a different hidden feature type for which the equivariance holds for all testing elements? Based on `rot3dOnR3.representations`, it would seem that the trivial feature type and `irrep` are my only options.

Thank you,

Jacob 

```
import escnn
import torch

class eqCNN(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.r3_act = escnn.gspaces.rot3dOnR3()
        self.feat_type_in = escnn.nn.FieldType(self.r3_act, [self.r3_act.irrep(1)])
        self.feat_type_hid = escnn.nn.FieldType(self.r3_act, 8 * [self.r3_act.trivial_repr])
        self.feat_type_out = escnn.nn.FieldType(self.r3_act, [self.r3_act.irrep(1)])
        self.model = escnn.nn.SequentialModule(
            escnn.nn.R3Conv(self.feat_type_in, self.feat_type_hid, kernel_size=3, padding=1),
            escnn.nn.InnerBatchNorm(self.feat_type_hid),
            escnn.nn.ReLU(self.feat_type_hid),
            escnn.nn.R3Conv(self.feat_type_hid, self.feat_type_hid, kernel_size=3, padding=1),
            escnn.nn.InnerBatchNorm(self.feat_type_hid),
            escnn.nn.ReLU(self.feat_type_hid),
            escnn.nn.R3Conv(self.feat_type_hid, self.feat_type_out, kernel_size=3, padding=1)
        )

    def forward(self, x):
        return self.model(escnn.nn.GeometricTensor(x, self.feat_type_in)).tensor

torch.manual_seed(1)
model = eqCNN()
x = torch.randn(1, 3, 8, 8, 8)

for j, g in enumerate(model.r3_act.testing_elements):
    x_transformed = escnn.nn.GeometricTensor(x, model.feat_type_in).transform(g).tensor
    y = model(x_transformed)
    y_transformed = escnn.nn.GeometricTensor(model(x), model.feat_type_in).transform(g).tensor
    if not torch.allclose(y, y_transformed, atol=1e-5):
        print(f"Failed for element {j}; mean error: {(y - y_transformed).abs().mean()}")
```

**Comment by Gabri95** [comment]
Hi @JacobHelwig 

The reason why the test works for the first 12 elements is that `rot3donR3.testing_elements` contains the 60 rotations of the Icosahedral group but, among these, only the first 12 are also symmetries of the cube, i.e. they only involve 90 or 180 deg rotations.

That means that the first 12 elements are also symmetries of the voxel grid containing your data.
However, the other 48 are not and will introduce some kind of interpolation artefacts when you rotate your input/outputs.

Testing for equivariance to rotations which don't preserve the grid should be done with some care to avoid (or at least reduce) these artefacts (e.g. work with higher resolution images such that you can rotate them at higher resolution and then downsample them. This should also be done for the outputs). 
Moreover, remember that perfect equivariance to these rotations is impossible so you always need to account for some equivariance error (up to 1e-2 or 1e-1 relative error)


Let me know if this answers your question!
Gabriele

**Comment by JacobHelwig** [comment]
Thank you, @Gabri95. This answers my question
